rest/access/handlers.go

```diff
@@ -46,7 +45,7 @@ func Authenticate(state *state.State, r *http.Request, trustedCerts map[string]x
 			}
 		}
 	default:
-		return false, fmt.Errorf("Invalid request address %q", r.Host)
```

Comment:
@masnax @roosterfish @tomponline I don't necessarily think that this change is a good idea. I'll keep this PR as draft until we've discussed in more detail.

`Authenticate` is called on every API request for all listeners. As microcluster is being used by more and more applications for various purposes, it's becoming a little too prescriptive in the way that requests are handled. 

In general, I think microcluster needs to be changed to remove the `AccessHandler` and `AllowUntrusted` from `EndpointAction` and instead accept a slice of handlers as middleware. Each handler should return a `response.Response`, if the response.Response is nil, we move onto the next handler. Downstreams can then implement their own authentication (and other middleware e.g. for logging/tracing). We can expose some handlers to downstreams that they may want to use - such as a wrapper for `Authenticate`.

Reply:
As for dropping `AllowUntrusted`, I'll give a bit of background as to how we treat trust in microcluster:

*  `Authenticate` is run by default using the default network address and the default truststore, and the trusted status of the request is stored in the request context's `CtxAccess`. This indicates whether the request is trusted by the core API and configuration. I think we should always keep a record of this for every endpoint in case we need to perform specific internal handling in the future. It will become very clear.

* Currently, if a downstream application wants to have their own access handler, they would have to set `AllowUntrusted=true` and create a custom `AccessHandler` that calls `Authenticate` with their specific parameters. In this case, the endpoint can be deemed trusted but the `CtxAccess` field in the request context will still be set to untrusted because the endpoint is not trusted according to the core configuration.

So EndpointActions behave like this:

```go
// Runs Authenticate with the core configuration, returns an error if untrusted.
rest.EndpointAction{Handler: endpointPost}

// Runs Authenticate with the core configuration, allows the request if trusted or untrusted.
rest.EndpointAction{Handler: endpointPost, AllowUntrusted: true}

// Runs myHandler, and if untrusted errors out. Then runs the core Authenticate, and if untrusted errors out.
rest.EndpointAction{Handler: endpointPost, AccessHandler: myHandler}

// Runs myHandler, and if untrusted errors out. Then runs the core Authenticate,  allows the request if untrusted.
rest.EndpointAction{Handler: endpointPost, AccessHandler: myHandler, AllowUntrusted: true}
```

I think it's useful to have `Authenticate` always run with the core configuration so we know whether an endpoint is trusted according to the default microcluster behaviour. 

If we get rid of the `AllowUntrusted` flag, that would require every endpoint to have an access handler. If an endpoint is supposed to be untrusted, we would need to define a no-op access handler that does not call `Authenticate` and also does not check `CtxAccess` populated by the internal `Authenticate` call. 

One thing I'm not sure of is what a list of access handlers would give us since you could just sequentially run the functions in an outer function anyway. It also makes it clearer to the writer that they need to either call a pre-defined access handler supplied by us or create their own, as opposed to specifying a list and digging through our code to see what the endpoint is doing with that list.

Reply:
> If we get rid of the `AllowUntrusted` flag, that would require every endpoint to have an access handler. If an endpoint is supposed to be untrusted, we would need to define a no-op access handler that does not call `Authenticate` and also does not check `CtxAccess` populated by the internal `Authenticate` call.

This is the same as the current `access.AllowAuthenticated` handler. I have a bone of contention with the naming of this handler because it doesn't actually check the value of `CtxAccess` and is a no-op. The naming assumes that the top level handler has already performed authentication.

> One thing I'm not sure of is what a list of access handlers would give us since you could just sequentially run the functions in an outer function anyway. It also makes it clearer to the writer that they need to either call a pre-defined access handler supplied by us or create their own, as opposed to specifying a list and digging through our code to see what the endpoint is doing with that list.

Not a list of access handlers, just a list of handlers. Microcluster is essentially a framework - this would allow easier creation of building blocks for composing sets of behaviour into the handlers. Information can be passed from one handler to the next via the request context. 

The core endpoints can follow the same pattern and continue to set `CtxAccess` in their own authentication middleware. Though I would additionally suggest that we expose these to be modified externally. As an example, as a user of microcloud, how do I change the authentication mechanism for core endpoints to use OIDC instead?

> ```go
> // Runs myHandler, and if untrusted errors out. Then runs the core Authenticate, and if untrusted errors out.
> rest.EndpointAction{Handler: endpointPost, AccessHandler: myHandler}
> 
> // Runs myHandler, and if untrusted errors out. Then runs the core Authenticate,  allows the request if untrusted.
> rest.EndpointAction{Handler: endpointPost, AccessHandler: myHandler, AllowUntrusted: true}
> ```

I don't think this is correct. Reading the code it looks like the top-level handler calls the core Authenticate function first, before calling the access handler. If we require that a user of microcluster sets `AllowUntrusted` to true and also set an `AccessHandler`, for all of these endpoint actions the call to `Authenticate` is superfluous.

Reply:
> > ```go
> > // Runs myHandler, and if untrusted errors out. Then runs the core Authenticate, and if untrusted errors out.
> > rest.EndpointAction{Handler: endpointPost, AccessHandler: myHandler}
> > 
> > // Runs myHandler, and if untrusted errors out. Then runs the core Authenticate,  allows the request if untrusted.
> > rest.EndpointAction{Handler: endpointPost, AccessHandler: myHandler, AllowUntrusted: true}
> > ```
> 
> I don't think this is correct. Reading the code it looks like the top-level handler calls the core Authenticate function first, before calling the access handler. 

Oops, yes I had it backwards but the idea remains which is that `AllowUntrusted` basically controls whether we care about the core Authenticate call. 

> If we require that a user of microcluster sets `AllowUntrusted` to true and also set an `AccessHandler`, for all of these endpoint actions the call to `Authenticate` is superfluous.

This is the only part I disagree with. I think it's mainly a naming issue. I think of `AllowUntrusted` as `RequireCoreAuthentication` (with the boolean flipped) and the supplied access handler as a second layer of access on top. So it's not superfluous, since they check different things. What's superfluous today is the needless `AllowAuthenticated` which does nothing.

I'm happy to drop the boolean, I'm just suggesting that we still keep the separation between internal/external authentication and always run the internal authentication. The "default" access handler would effectively serve the same purpose as `AllowUntrusted` does today by actually checking the value of what the internal authentication determined.

Reply:
I've added a `hostAddress` argument to the `Authenticate` method so that it can be used in other contexts. The method now returns an error of type `access.ErrInvalidHost` if the host does not match the expected value. In the top-level handler, errors of this type are ignored so that the request can continue (the request is still considered untrusted from a core API perspective because the value of `trusted` in the request context is still false).

The `AllowAuthenticated` access handler is now always called when `AllowUntrusted` is false. This handler now checks the the `trusted` value in the request context.

This means that when running extra listeners via microcloud, downstreams must set `AllowUntrusted` to true on all endpoint actions and implement their own access handlers.